EIPS/eip-165.md

```diff
@@ -0,0 +1,231 @@
+## Motivation
+
+For some "standard interfaces" like the ERC 20 token interface ( [#20](https://github.com/ethereum/EIPs/issues/20) ), it is sometimes useful to query whether a contract supports the interface and if yes, which version of the interface, in order to adapt the way in which the contract is to be interfaced with. Specifically for #20, a version identifier has already been proposed. This proposal wants to standarize the concept of interfaces and map interface versions to interface identifiers.
+
+## Specification
+##### Standard interface identifier.
+
+Any standard interface will be identified by a specific `bytes4` number. This number will be the XOR of all `bytes4` of each method calculated as solidity does.
+
+For example, to calculate the id of a standard interface
+
+totalSupply()               -> Method ID: 0x18160ddd
+balanceOf(address)          -> Method ID: 0x70a08231
+transfer(address, uint256)  -> Method ID: 0xa9059cbb
+
+InterfaceID = 0x18160ddd ^ 0x70a08231 ^ 0xa9059cbb = 0xc1b31357
+
+##### EIP165 Interface
+
+Any contract that wants to implement a standard interface must also implement this method:
+
+    function supportsInterface(bytes4 interfaceID) returns (bool);
+
+This function must return:
+    `true` when `interfaceID` is 0x01ffc9a7 (EIP165 interface)
+    `false` when `interfaceID` is 0xffffffff
+    `true` for any `interfaceID` that this contract implements.
+    `false` for any other `interfaceID`
+
+##### Procedure to determine if a contract implements EIP165 standard.
+
+1. The source contact makes a `CALL` to the destination address with input data: 0x01ffc9a701ffc9a7 value: 0 and gas 30000
+2. If the call fails or return false, the destination contract does not implement EIP165
+3. If the call returns true, a second call is made with input data: 0x01ffc9a7ffffffff
+4. If the call fails or returns true, the destination contract does not implement EIP165
+5. Otherwise it implements EIP165
+
+##### Procedure to determine if a contract implements a standard interface
+
+1. If you are not sure if the contract implements EIP165 Interface, use the previous procedure to confirm.
+2. If it does not implement EIP165, then you will have to see what methods it uses the old fashioned way.
+3. If it implements EIP165 then just call `supportsInterface(interfaceID)` to determine if it implements an interface you can use.
+
+##### EIP165Cache contract.
+
+The contract on address 0x_________ can also be used to determine if a contract implements EIP165, an interface, or a set of interfaces. This contract will use the previous procedure to determine the implementation of those interfaces and will hold a cache to optimize the gas costs of those queries.
+
+This contract will implement the next functions:
+
+    function eip165Supported(address _contract) constant returns (bool);
+
+Returns true if `_contract` supports EIP165
+
+    function interfaceSupported(address _contract, bytes4 _interfaceId) constant returns (bool);
```

Comment:
If the contract `throws` (because it is an old one) it will consume all the gas.... So in this case I suspect that it's much more efficient to have a `cache`.  For a new contract (and if we remove the double call), probably it will not be very different.

Reply:
Good point on the `throw`.

Reply:
If we assume that we will use a cache, then may be it's not so important to remove the "double call" in terms of performance (It will be done only once by the cache contract). And this way we don't risk for edge cases.

Reply:
True; how about rewording it such that it's optional? In cases where you already know the target contract implements `supportsInterface`, there's no point in calling it more than once.